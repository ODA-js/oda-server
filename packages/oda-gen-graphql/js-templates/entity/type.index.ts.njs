<#@ context 'entity' -#>
<#@ requireAs ('entity/type/enums.graphql.njs', 'enums') #>
<#@ requireAs ('entity/type/entry.graphql.njs', 'type') #>
<#@ requireAs ('entity/connections/types.graphql.njs', 'connections.types') #>
<#@ requireAs ('entity/connections/mutations/types.graphql.njs', 'connections.mutation') #>
<#@ requireAs ('entity/connections/mutations/entry.graphql.njs', 'connections.mutation.entry') #>
<#@ requireAs ('entity/mutations/types.graphql.njs', 'mutation.types') #>
<#@ requireAs ('entity/mutations/entry.graphql.njs', 'mutation.entry') #>
<#@ requireAs ('entity/subscriptions/types.graphql.njs', 'subscription.types') #>
<#@ requireAs ('entity/subscriptions/entry.graphql.njs', 'subscription.entry') #>
<#@ requireAs ('entity/query/entry.graphql.njs', 'query.entry') #>
<#@ requireAs ('entity/viewer/entry.graphql.njs', 'viewer.entry') #>

import { common } from 'oda-gen-graphql';

let { fillDefaults, deepMerge } = common.lib;

import { query } from './query/resolver';
import { viewer } from './viewer/resolver';
import { resolver } from './type/resolver';
import { mutation as connectionMutation } from './connections/mutations/resolver';
import { mutation as entityMutation } from './mutations/resolver';
import { subscriptions as entitySubscription, resolver as subscriptionsUnions } from './subscriptions/resolver';

export class #{entity.name} extends common.types.GQLModule {
  constructor(_args) {
    super(_args);
    this._name = '#{entity.name}';
    this._query = fillDefaults(this._query, query);
    this._viewer = fillDefaults(this._viewer, viewer);
    this._resolver = fillDefaults(this._resolver, resolver, subscriptionsUnions);

    this._typeDef = fillDefaults(this._typeDef, {
      'enums': [`#{partial(entity.partials['enums'], 'enums')}`],
      'type': [`#{partial(entity.partials['type'], 'type')}`],
      'mutationTypes': [`#{partial(entity.partials['mutation.types'], 'mutation.types')}`],
      'subscriptionsTypes': [`#{partial(entity.partials['subscription.types'], 'subscription.types')}`],
      'connectionsTypes': [`#{partial(entity.partials['connections.types'], 'connections.types')}`],
      'connectionsMutation': [`#{partial(entity.partials['connections.mutation'], 'connections.mutation')}`],
    });

    this._mutationEntry = fillDefaults(this._mutationEntry, {
      'mutationEntry': [`#{partial(entity.partials['mutation.entry'], 'mutation.entry')}`],
      'connectionsMutationEntry': [`#{partial(entity.partials['connections.mutation.entry'],'connections.mutation.entry')}`],
    });

    this._subscriptionEntry = fillDefaults(this._subscriptionEntry, {
      'subscriptionEntry': [`#{partial(entity.partials['subscription.entry'], 'subscription.entry')}`],
    });

    this._queryEntry = fillDefaults(this._queryEntry, {
      'queryEntry': [`#{partial(entity.partials['query.entry'], 'query.entry')}`],
    });

    this._mutation = fillDefaults(this._mutation, deepMerge(entityMutation, connectionMutation));
    this._subscription = fillDefaults(this._subscription, entitySubscription);

    this._viewerEntry = fillDefaults(this._viewerEntry, {
      'viewerEntry': [`#{partial(entity.partials['viewer.entry'], 'viewer.entry')}`],
    });
  }
}
