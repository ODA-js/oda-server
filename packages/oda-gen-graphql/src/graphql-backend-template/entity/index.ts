import * as connections from './connections';
import * as data from './data';
import * as mutations from './mutations';
import * as dataPump from './dataPump';
import * as subscriptions from './subscriptions';
import * as query from './query';
import * as viewer from './viewer';
import * as type from './type';
import * as index from './type.index.ts';
import * as UI from './UI';

export {
  connections,
  data,
  mutations,
  dataPump,
  subscriptions,
  query,
  viewer,
  type,
  index,
  UI,
};
