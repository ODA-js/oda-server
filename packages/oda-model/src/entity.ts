import {
  EntityBasePersistence,
  EntityBaseInternal,
  EntityBaseMetaInfo,
  EntityBaseInput,
  IEntityBase,
  EntityBase,
  EntityBaseOutput,
} from './entitybase';
import { MetaModelType, Nullable, assignValue } from './types';
import { merge } from 'lodash';

export interface IEntity
  extends IEntityBase<
    EntityMetaInfo,
    EntityPersistence,
    EntityInput,
    EntityOutput
  > {
  implements: Set<string>;
  embedded: boolean | Set<string>;
  abstract: boolean;
}

export interface EntityInternal
  extends EntityBaseInternal<EntityMetaInfo, EntityPersistence> {
  implements: Set<string>;
  embedded: boolean | Set<string>;
  abstract: boolean;
}

export interface EntityPersistence extends EntityBasePersistence {}

export interface EntityInput
  extends EntityBaseInput<EntityMetaInfo, EntityPersistence> {
  implements?: string[];
  embedded?: boolean | string[];
  abstract?: boolean;
}

export interface EntityOutput
  extends EntityBaseOutput<EntityMetaInfo, EntityPersistence> {
  implements?: string[];
  embedded?: boolean | string[];
  abstract?: boolean;
}

export interface EntityMetaInfo extends EntityBaseMetaInfo<EntityPersistence> {}

const defaultMetaInfo = {};
const defaultInput = { metadata: defaultMetaInfo };

export class Entity
  extends EntityBase<
    EntityMetaInfo,
    EntityInput,
    EntityInternal,
    EntityPersistence,
    EntityOutput
  >
  implements IEntity {
  public get modelType(): MetaModelType {
    return 'entity';
  }

  constructor(init: EntityInput) {
    super(merge({}, defaultInput, init));
  }

  get implements(): Set<string> {
    return this.$obj.implements;
  }

  get abstract(): boolean {
    return this.$obj.abstract;
  }

  get embedded(): boolean | Set<string> {
    return this.$obj.embedded;
  }

  public updateWith(input: Nullable<EntityInput>) {
    super.updateWith(input);

    assignValue({
      src: this.$obj,
      input,
      field: 'abstract',
      setDefault: src => (src.abstract = false),
    });
    assignValue<EntityInternal, EntityInput, string[]>({
      src: this.$obj,
      input,
      field: 'implements',
      effect: (src, value) => (src.implements = new Set(value)),
      setDefault: src => (src.implements = new Set()),
    });

    assignValue<EntityInternal, EntityInput, boolean | string[]>({
      src: this.$obj,
      input,
      field: 'embedded',
      effect: (src, value) =>
        (src.embedded = typeof value === 'boolean' ? value : new Set(value)),
      setDefault: src => (src.embedded = false),
    });
  }

  public toObject(): EntityOutput {
    return merge({}, super.toObject(), {
      implements: [...this.implements.values()],
      embedded: this.embedded,
      abstract: this.abstract,
    } as Partial<EntityOutput>);
  }
}
