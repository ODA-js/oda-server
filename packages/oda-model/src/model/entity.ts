import clean from '../lib/json/clean';
import { Field } from './field';
import {
  EntityInput,
  EntityJSON,
  EntityStorage,
  IEntity,
  MetaModelType,
} from './interfaces';
import { ModelPackage } from './modelpackage';
import { EntityBase } from './entitybase';

/**
 * 1. тип объекта который входит на updateWith
 * 2. тип объекта который идет на toObject
 * 3. тип объекта который идет на toJSON
 * 3. тип объекта который идет на выходе clone
 */

export class Entity extends EntityBase implements IEntity {
  public modelType: MetaModelType = 'entity';

  protected $obj!: EntityStorage;

  constructor(obj: EntityInput) {
    super(obj);
  }

  get implements(): Set<string> {
    return this.$obj.implements;
  }

  get abstract(): boolean {
    return !!this.$obj.abstract;
  }

  get embedded(): boolean | string[] {
    return this.$obj.embedded instanceof Set
      ? Array.from(this.$obj.embedded)
      : this.$obj.embedded;
  }

  public ensureImplementation(modelPackage: ModelPackage) {
    const newFields: Map<string, Field> = new Map<string, Field>();
    this.implements.forEach(intrf => {
      if (modelPackage.mixins.has(intrf)) {
        const impl = modelPackage.mixins.get(intrf);
        impl.fields.forEach(f => {
          if (!this.fields.has(f.name)) {
            newFields.set(f.name, f);
          }
        });
      }
    });

    if (newFields.size > 0) {
      const update = this.toJSON();
      update.fields.push(...[...newFields.values()].map(f => f.toJSON()));
      this.updateWith(update);
      this.ensureIds(modelPackage);
    }
  }

  public updateWith(obj: EntityInput) {
    if (obj) {
      super.updateWith(obj);

      const result = { ...this.$obj };
      const impl = new Set(obj.implements);
      result.embedded = Array.isArray(obj.embedded)
        ? new Set(obj.embedded)
        : obj.embedded;
      result.abstract = obj.abstract;
      result.implements = impl;
      this.$obj = result;
    }
  }

  public toObject(modelPackage?: ModelPackage) {
    if (!modelPackage) {
      let res = super.toObject();
      return clean({
        ...res,
        implements: [...this.implements],
        embedded: this.embedded,
        abstract: this.abstract,
      });
    } else {
      let modelRelations = modelPackage.relations.get(this.name);
      if (modelRelations) {
        let res = super.toObject();
        return clean({
          ...res,
          embedded: this.embedded,
          abstract: this.abstract,
          implements: [...this.implements].filter(i =>
            modelPackage.mixins.has(i),
          ),
        });
      }
    }
  }
}
