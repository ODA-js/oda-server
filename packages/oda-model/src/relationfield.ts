import { merge } from 'lodash';
import { Nullable, assignValue, MetaModelType } from './types';
import { RelationInput } from './relation';
import { HasOne } from './hasone';
import { HasMany } from './hasmany';
import { BelongsToMany } from './belongstomany';
import { BelongsTo } from './belongsto';
import {
  RelationFieldBase,
  RelationFieldBaseMetaInfo,
  RelationFieldBasePersistence,
  IRelationFieldBase,
  RelationFieldBaseInternal,
  RelationFieldBaseInput,
  RelationFieldBaseOutput,
} from './relationfieldbase';
import { Internal } from './element';

function discoverFieldType(obj: any) {
  // сделать проверку по полю...
  if (obj.hasOne) {
    return 'HasOne';
  } else if (obj.hasMany) {
    return 'HasMany';
  } else if (obj.belongsTo) {
    return 'BelongsTo';
  } else if (obj.belongsToMany) {
    return 'BelongsToMany';
  } else {
    console.warn(`undefined relation type of ${JSON.stringify(obj)}`);
    return 'undefined';
  }
}

export interface RelationFieldMetaInfo
  extends RelationFieldBaseMetaInfo<RelationFieldPersistence> {}

export interface RelationFieldPersistence
  extends RelationFieldBasePersistence {}

export interface RelationFieldInput
  extends RelationFieldBaseInput<
    RelationFieldMetaInfo,
    RelationFieldPersistence
  > {
  relation: RelationInput;
}

export interface RelationFieldOutput
  extends RelationFieldBaseOutput<
    RelationFieldMetaInfo,
    RelationFieldPersistence
  > {
  relation: RelationInput;
}

/**
 * relation field definition
 */
export interface IRelationField
  extends IRelationFieldBase<
    RelationFieldMetaInfo,
    RelationFieldInput,
    RelationFieldPersistence,
    RelationFieldOutput
  > {}

export interface RelationFieldInput
  extends RelationFieldBaseInput<
    RelationFieldMetaInfo,
    RelationFieldPersistence
  > {}

export interface RelationFieldInternal extends RelationFieldBaseInternal {}

const defaultMetaInfo = {};
const defaultInput = {
  metadata: defaultMetaInfo,
  persistent: false,
  derived: false,
  identity: false,
  indexed: false,
  required: false,
};

export class RelationField
  extends RelationFieldBase<
    RelationFieldMetaInfo,
    RelationFieldInput,
    RelationFieldInternal,
    RelationFieldPersistence,
    RelationFieldOutput
  >
  implements IRelationField {
  public get modelType(): MetaModelType {
    return 'relation-field';
  }
  constructor(init: RelationFieldInput) {
    super(merge({}, defaultInput, init));
  }

  public updateWith(input: Nullable<RelationFieldInput>) {
    super.updateWith(input);

    assignValue<
      RelationFieldInternal,
      RelationFieldInput,
      RelationFieldInput['relation']
    >({
      src: this[Internal],
      input,
      field: 'relation',
      effect: (src, value) => {
        switch (discoverFieldType(value)) {
          case 'HasOne':
            src.relation = new HasOne({
              ...(value as { hasOne: string }),
              entity: this.metadata.entity,
              field: this.name,
            });
            break;
          case 'HasMany':
            src.relation = new HasMany({
              ...(value as { hasMany: string }),
              entity: this.metadata.entity,
              field: this.name,
            });
            break;
          case 'BelongsToMany':
            src.relation = new BelongsToMany({
              ...(value as { belongsToMany: string; using: string }),
              entity: this.metadata.entity,
              field: this.name,
            });
            break;
          case 'BelongsTo':
            src.relation = new BelongsTo({
              ...(value as { belongsTo: string }),
              entity: this.metadata.entity,
              field: this.name,
            });
            break;
          default:
            throw new Error('undefined type');
        }
      },
    });

    assignValue<
      RelationFieldMetaInfo,
      RelationFieldInput,
      NonNullable<RelationFieldInput['derived']>
    >({
      src: this.metadata,
      input,
      inputField: 'derived',
      effect: (src, value) => {
        if (
          this.relation.verb === 'HasOne' ||
          this.relation.verb === 'HasMany'
        ) {
          src.persistence.derived = value;
        } else {
          src.persistence.derived = false;
        }
      },
      required: true,
      setDefault: src => (src.persistence.derived = defaultInput.derived),
    });

    assignValue<
      RelationFieldMetaInfo,
      RelationFieldInput,
      NonNullable<RelationFieldInput['persistent']>
    >({
      src: this.metadata,
      input,
      inputField: 'persistent',
      effect: (src, value) => {
        if (this.relation.verb === 'BelongsTo') {
          src.persistence.persistent = value;
        } else {
          src.persistence.persistent = false;
        }
      },
      required: true,
      setDefault: src => (src.persistence.persistent = defaultInput.persistent),
    });

    assignValue<
      RelationFieldMetaInfo,
      RelationFieldInput,
      NonNullable<RelationFieldInput['identity']>
    >({
      src: this.metadata,
      input,
      inputField: 'identity',
      effect: (src, value) => {
        if (this.relation.verb === 'BelongsTo') {
          src.persistence.identity = value;
        } else {
          src.persistence.identity = false;
        }
      },
      required: true,
      setDefault: src => (src.persistence.identity = defaultInput.identity),
    });

    assignValue<
      RelationFieldMetaInfo,
      RelationFieldInput,
      NonNullable<RelationFieldInput['indexed']>
    >({
      src: this.metadata,
      input,
      inputField: 'indexed',
      effect: (src, value) => {
        if (this.relation.verb === 'BelongsTo') {
          src.persistence.indexed = value;
        } else {
          src.persistence.indexed = false;
        }
      },
      required: true,
      setDefault: src => (src.persistence.indexed = defaultInput.indexed),
    });

    assignValue<
      RelationFieldMetaInfo,
      RelationFieldInput,
      NonNullable<RelationFieldInput['required']>
    >({
      src: this.metadata,
      input,
      inputField: 'required',
      effect: (src, value) => {
        if (this.relation.verb === 'BelongsTo') {
          src.persistence.required = value;
        } else {
          src.persistence.required = false;
        }
      },
      required: true,
      setDefault: src => (src.persistence.required = defaultInput.required),
    });
  }

  // it get fixed object
  public toObject(): RelationFieldOutput {
    return merge({}, super.toObject(), {
      relation: this.relation.toObject(),
    } as Partial<RelationFieldOutput>);
  }

  public mergeWith(payload: Nullable<RelationFieldInput>) {
    super.mergeWith(payload);
  }
}
