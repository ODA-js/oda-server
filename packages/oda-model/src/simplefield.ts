import {
  FieldBaseInput,
  FieldBaseInternal,
  FieldBaseMetaInfo,
  IFieldBase,
  FieldBase,
  FieldBasePersistence,
  FieldBaseOutput,
} from './fieldbase';
import { merge, get } from 'lodash';
import { EnumType, Nullable, assignValue, MetaModelType } from './types';
import { Internal } from './element';
``;
export interface ISimpleField
  extends IFieldBase<
    SimpleFieldMeta,
    SimpleFieldInput,
    SimpleFieldPersistence,
    SimpleFieldOutput
  > {
  readonly type: string | EnumType;
  readonly list?: boolean;
  readonly defaultValue?: string;
}

export interface SimpleFieldPersistence extends FieldBasePersistence {
  derived: boolean;
  persistent: boolean;
}

export interface SimpleFieldMeta
  extends FieldBaseMetaInfo<SimpleFieldPersistence> {
  defaultValue?: string;
}

export interface SimpleFieldInternal extends FieldBaseInternal {
  type: string | EnumType;
  list: boolean;
}

export interface SimpleFieldInput
  extends FieldBaseInput<SimpleFieldMeta, SimpleFieldPersistence> {
  type?: string | EnumType;
  defaultValue?: string;
  list?: boolean;
}

export interface SimpleFieldOutput
  extends FieldBaseOutput<SimpleFieldMeta, SimpleFieldPersistence> {
  type?: string | EnumType;
  defaultValue?: string;
  list?: boolean;
}

export const simpleFieldDefaultMetaInfo = {};
export const simpleFieldDefaultInput = { metadata: simpleFieldDefaultMetaInfo };

export class SimpleField
  extends FieldBase<
    SimpleFieldMeta,
    SimpleFieldInput,
    SimpleFieldInternal,
    SimpleFieldPersistence,
    SimpleFieldOutput
  >
  implements ISimpleField {
  public get modelType(): MetaModelType {
    return typeof this[Internal].type === 'string'
      ? 'simple-field'
      : 'enum-field';
  }

  get type(): string | EnumType {
    return this[Internal].type;
  }

  // if it is the field is List of Items i.e. String[]
  get list(): boolean {
    return get(this[Internal], 'list', false);
  }

  get defaultValue(): string | undefined {
    return get(this.metadata, 'defaultValue');
  }

  constructor(init: SimpleFieldInput) {
    super(merge({}, simpleFieldDefaultInput, init));
  }

  public updateWith(input: Nullable<SimpleFieldInput>) {
    super.updateWith(input);

    assignValue<
      SimpleFieldInternal,
      SimpleFieldInput,
      NonNullable<SimpleFieldInput['type']>
    >({
      src: this[Internal],
      input,
      field: 'type',
      effect: (src, value) => {
        if (typeof value !== 'string' && !value.multiplicity) {
          value.multiplicity = 'one';
        }
        src.type = value;
      },
      setDefault: src => (src.type = 'string'),
    });

    assignValue<
      SimpleFieldInternal,
      SimpleFieldInput,
      NonNullable<SimpleFieldInput['list']>
    >({
      src: this[Internal],
      input,
      field: 'list',
      required: true,
      effect: (src, value) => {
        if (this.modelType === 'enum-field') {
          (src.type as EnumType).multiplicity = value ? 'many' : 'one';
        }
        src.list = value;
      },
      setDefault: src => {
        if (this.modelType === 'enum-field') {
          src.list =
            (src.type as EnumType).multiplicity === 'many' ? true : false;
        } else {
          src.list = false;
        }
      },
    });

    assignValue<
      SimpleFieldMeta,
      SimpleFieldInput,
      NonNullable<SimpleFieldInput['derived']>
    >({
      src: this.metadata,
      input,
      inputField: 'derived',
      effect: (src, value) => {
        src.persistence.derived = value;
        if (!value) {
          this.metadata.persistence.persistent = true;
        }
      },
      setDefault: src => (src.persistence.derived = false),
    });

    assignValue<
      SimpleFieldMeta,
      SimpleFieldInput,
      NonNullable<SimpleFieldInput['persistent']>
    >({
      src: this.metadata,
      input,
      inputField: 'persistent',
      effect: (src, value) => (src.persistence.persistent = value),
      setDefault: src => (src.persistence.persistent = true),
    });

    assignValue<
      SimpleFieldMeta,
      SimpleFieldInput,
      NonNullable<SimpleFieldInput['args']>
    >({
      src: this.metadata,
      input,
      inputField: 'args',
      allowEffect: (_, value) =>
        Array.isArray(value) ? value.length > 0 : Object.keys(value).length > 0,
      effect: (src, _) => (src.persistence.derived = true),
    });

    assignValue<
      SimpleFieldMeta,
      SimpleFieldInput,
      NonNullable<SimpleFieldInput['defaultValue']>
    >({
      src: this.metadata,
      input,
      inputField: 'defaultValue',
      effect: (src, value) => {
        src.defaultValue = value;
      },
      allowEffect: (src, _) => !src.persistence.derived,
    });
  }

  public toObject(): SimpleFieldOutput {
    return merge({}, super.toObject(), {
      type: this.type,
      inheritedFrom: this[Internal].inheritedFrom,
      list: this[Internal].list,
    } as Partial<SimpleFieldOutput>);
  }

  public mergeWith(payload: Nullable<SimpleFieldInput>) {
    super.mergeWith(payload);
  }
}
