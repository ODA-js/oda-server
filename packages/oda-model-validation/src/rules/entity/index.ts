import pluralName from './pluralName';

export default [new pluralName()];
