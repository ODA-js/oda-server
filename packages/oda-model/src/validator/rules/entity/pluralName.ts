import { IValidationResult } from '../../../interfaces';
import { IEntityContext } from '../../interfaces';
import { Rule } from '../../rules';

export default class implements Rule<IEntityContext> {
  public name = 'entity-plural-name-the-same';
  public description =
    'plural form of entity`s name of entity must be different from its singular form';
  public validate(context: IEntityContext): IValidationResult[] {
    const result: IValidationResult[] = [];
    if (context.entity.name === context.entity.plural) {
      result.push({
        entity: context.entity.name,
        message: this.description,
        result: 'error',
      });
    }
    return result;
  }
}
