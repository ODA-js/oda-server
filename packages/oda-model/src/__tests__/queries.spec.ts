import 'jest';
import {
  idField,
  uniqueField,
  mutableFields,
  storedRelations,
  getGeneralIndexedFields,
  getUniqueIndexedFields,
  ArgsFromTuples,
} from '../utils/queries';
import { SimpleField } from '../simplefield';
import { RelationField } from '../relationfield';
import { Entity } from '../entity';

describe('queries', () => {
  it('idField', () => {
    expect(
      idField(new SimpleField({ name: '_id', identity: true })),
    ).toBeTruthy();
    expect(
      idField(new SimpleField({ name: 'id', identity: true })),
    ).toBeTruthy();
  });
  it('uniqueField', () => {
    expect(
      uniqueField(new SimpleField({ name: 'abc', identity: true })),
    ).toBeTruthy();
  });
  it('mutableFields', () => {
    expect(
      mutableFields(
        new SimpleField({ name: 'abc', identity: true, persistent: true }),
      ),
    ).toBeTruthy();
  });
  it('storedRelations', () => {
    expect(
      storedRelations(
        new RelationField({
          name: 'PD',
          relation: {
            belongsTo: 'D#',
          },
        }),
      ),
    ).toBeTruthy();
  });
  it('indexes', () => {
    const entity = new Entity({
      name: 'ABC',
      fields: [
        {
          name: 'a',
          indexed: true,
        },
        {
          name: 'b',
          indexed: 'b',
        },
        {
          name: 'c',
          indexed: 'b',
        },
      ],
    });
    expect(
      ArgsFromTuples(getGeneralIndexedFields(entity), name => name),
    ).toMatchSnapshot('all ndexes');
  });
  it('unique indexes', () => {
    const entity = new Entity({
      name: 'ABC',
      fields: [
        {
          name: 'a',
          identity: true,
        },
        {
          name: 'b',
          identity: 'b',
        },
        {
          name: 'c',
          identity: 'b',
        },
      ],
    });
    expect(
      ArgsFromTuples(getUniqueIndexedFields(entity), name => name),
    ).toMatchSnapshot('all identity indexes');
  });
});
