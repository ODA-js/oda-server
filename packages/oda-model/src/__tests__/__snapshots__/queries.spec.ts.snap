// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries indexes: all ndexes 1`] = `
Array [
  Object {
    "description": undefined,
    "kind": "input",
    "name": "a",
    "order": 0,
    "required": false,
    "title": undefined,
    "type": "string",
  },
  Object {
    "fields": Array [
      Object {
        "description": undefined,
        "kind": "input",
        "name": "b",
        "order": 0,
        "required": false,
        "title": undefined,
        "type": "string",
      },
      Object {
        "description": undefined,
        "kind": "input",
        "name": "c",
        "order": 1,
        "required": false,
        "title": undefined,
        "type": "string",
      },
    ],
    "name": "b",
  },
]
`;

exports[`queries unique indexes: all identity indexes 1`] = `
Array [
  Object {
    "description": undefined,
    "kind": "input",
    "name": "a",
    "order": 0,
    "required": true,
    "title": undefined,
    "type": "string",
  },
  Object {
    "fields": Array [
      Object {
        "description": undefined,
        "kind": "input",
        "name": "b",
        "order": 0,
        "required": true,
        "title": undefined,
        "type": "string",
      },
      Object {
        "description": undefined,
        "kind": "input",
        "name": "c",
        "order": 1,
        "required": true,
        "title": undefined,
        "type": "string",
      },
    ],
    "name": "b",
  },
  Object {
    "description": undefined,
    "kind": "input",
    "name": "id",
    "order": 0,
    "required": true,
    "title": undefined,
    "type": "ID",
  },
]
`;
